src/main/kotlin/uk/gov/justice/digital/hmpps/courthearingeventreceiver/service/MessageNotifier.kt

```diff
@@ -34,10 +48,49 @@ class MessageNotifier(
         .stringValue(hearingEventType.description)
         .build()
 
+    if (sizeOfMessage(hearingEvent) >= maxMessageSize) {
+      publishLargeMessage(hearingEventType = hearingEventType, hearingEvent = hearingEvent)
+      return
+    }
+
     val publishResult = topic.publish(eventType = "commonplatform.case.received", event = objectMapper.writeValueAsString(hearingEvent), attributes = mapOf("messageType" to messageTypeValue, "hearingEventType" to hearingEventTypeValue), messageGroupId = MESSAGE_GROUP_ID)
+
     log.info("Published message with message Id {}", publishResult.messageId())
   }
 
+  fun publishLargeMessage(hearingEventType: HearingEventType, hearingEvent: HearingEvent) {
+    val hearingEventTypeValue =
+      MessageAttributeValue.builder()
+        .dataType("String")
+        .stringValue(hearingEventType.description)
+        .build()
+
+    // hmpps.topic library returns SNSAsyncClient which requires S3AsyncClient
+    val snsExtendedAsyncClientConfiguration: SNSExtendedAsyncClientConfiguration = SNSExtendedAsyncClientConfiguration()
+      .withPayloadSupportEnabled(amazonS3AsyncClient, bucketName)
+
+    val snsExtendedClient = AmazonSNSExtendedAsyncClient(
+      topic.snsClient,
+      snsExtendedAsyncClientConfiguration,
+    )
+    // Publish message via SNS with storage in S3
+    val publishResult = snsExtendedClient.publish(
+      PublishRequest.builder().topicArn(topic.arn).messageAttributes(
+        mapOf(
+          "messageType" to MessageAttributeValue.builder().dataType("String").stringValue(MESSAGE_TYPE).build(),
+          "hearingEventType" to hearingEventTypeValue,
+          "eventType" to MessageAttributeValue.builder().dataType("String").stringValue("commonplatform.case.received").build(),
+        ),
+      ).messageGroupId(MESSAGE_GROUP_ID).message(objectMapper.writeValueAsString(hearingEvent))
+        .build(),
+    )
+    log.info("Published large message with message Id {}", publishResult.get().messageId())
+  }
+
+  fun sizeOfMessage(hearingEvent: HearingEvent): Int {
+    return hearingEvent.toString().toByteArray().size
```

Comment:
A minor point but I would be inclined to do something like this for readability, rather than an `if` with a return statement

```
when {
  message.isLarge() -> publishLargeMessage()
  else -> publishAsUsual()
}
```

Reply:
cool solution, i like it 👍🏾